substrate/primitives/runtime/src/generic/checked_extrinsic.rs

```diff
@@ -18,81 +18,124 @@
 		source: TransactionSource,
 		info: &DispatchInfoOf<Self::Call>,
 		len: usize,
 	) -> TransactionValidity {
-		if let Some((ref id, ref extra)) = self.signed {
-			Extra::validate(extra, id, &self.function, info, len)
-		} else {
-			let valid = Extra::validate_unsigned(&self.function, info, len)?;
-			let unsigned_validation = U::validate_unsigned(source, &self.function)?;
-			Ok(valid.combine_with(unsigned_validation))
+		match self.format {
+			ExtrinsicFormat::Bare => {
+				let inherent_validation = I::validate_unsigned(source, &self.function)?;
+				#[allow(deprecated)]
+				let legacy_validation = Extension::validate_bare_compat(&self.function, info, len)?;
+				Ok(legacy_validation.combine_with(inherent_validation))
+			},
+			ExtrinsicFormat::Signed(ref signer, ref extension) => {
+				let origin = Some(signer.clone()).into();
+				extension.validate_only(origin, &self.function, info, len).map(|x| x.0)
+			},
+			ExtrinsicFormat::General(ref extension) =>
+				extension.validate_only(None.into(), &self.function, info, len).map(|x| x.0),
 		}
 	}
 
-	fn apply<U: ValidateUnsigned<Call = Self::Call>>(
+	fn apply<I: ValidateUnsigned<Call = Self::Call>>(
 		self,
 		info: &DispatchInfoOf<Self::Call>,
 		len: usize,
 	) -> crate::ApplyExtrinsicResultWithInfo<PostDispatchInfoOf<Self::Call>> {
-		let (maybe_who, maybe_pre) = if let Some((id, extra)) = self.signed {
-			let pre = Extra::pre_dispatch(extra, &id, &self.function, info, len)?;
-			(Some(id), Some(pre))
-		} else {
-			Extra::pre_dispatch_unsigned(&self.function, info, len)?;
-			U::pre_dispatch(&self.function)?;
-			(None, None)
-		};
-		let res = self.function.dispatch(RuntimeOrigin::from(maybe_who));
-		let post_info = match res {
-			Ok(info) => info,
-			Err(err) => err.post_info,
-		};
-		Extra::post_dispatch(
-			maybe_pre,
-			info,
-			&post_info,
-			len,
-			&res.map(|_| ()).map_err(|e| e.error),
-		)?;
-		Ok(res)
+		match self.format {
+			ExtrinsicFormat::Bare => {
+				I::pre_dispatch(&self.function)?;
+				// TODO: Remove below once `ValidateUnsigned` is deprecated and the only `Bare`
+				// extrinsics left are inherents.
+				#[allow(deprecated)]
+				Extension::validate_bare_compat(&self.function, info, len)?;
+				#[allow(deprecated)]
+				Extension::pre_dispatch_bare_compat(&self.function, info, len)?;
+				let res = self.function.dispatch(None.into());
+				let post_info = res.unwrap_or_else(|err| err.post_info);
+				let pd_res = res.map(|_| ()).map_err(|e| e.error);
+				// TODO: Remove below once `ValidateUnsigned` is deprecated and the only `Bare`
+				// extrinsics left are inherents.
+				#[allow(deprecated)]
+				Extension::post_dispatch_bare_compat(info, &post_info, len, &pd_res)?;
+				Ok(res)
+			},
+			ExtrinsicFormat::Signed(signer, extension) =>
+				extension.dispatch_transaction(Some(signer).into(), self.function, info, len),
+			ExtrinsicFormat::General(extension) =>
+				extension.dispatch_transaction(None.into(), self.function, info, len),
```

Comment:
In light of recent discussions, I think we should deny the origin `None` after the transaction extension `validate` is executed.

Reply:
Done.

Reply:
code is good to me